lib/improver/spotdata/neighbour_finding.py

```diff
@@ -0,0 +1,445 @@
+#
+#         neighbours['i'] = i_indices
+#         neighbours['j'] = j_indices
+#         neighbours['dz'] = dz
+#
+#         return neighbours
+
+    def minimum_height_error_neighbour(self, cube, sites,
+                                       default_neighbours=None,
+                                       relative_z=None,
+                                       land_constraint=False,
+                                       ancillary_data=None):
+
+        '''
+        Find the horizontally nearest neighbour, then relax the conditions
+        to find the neighbouring point in the 9 nearest nodes to the input
+        coordinate that minimises the height difference. This is typically
+        used for temperature, where vertical displacement can be much more
+        important that horizontal displacement in determining the conditions.
+
+        A vertical displacement bias may be applied with the relative_z
+        keyword; whether to prefer grid points above or below the site, or
+        neither.
+
+        A land constraint may be applied that requires a land grid point be
+        selected for a site that is over land. Currently this is established
+        by checking that the nearest grid point barring any other conditions
+        is a land point. If a site is a sea point it will use the nearest
+        neighbour as there should be no vertical displacement difference with
+        other sea points.
+
+        Args:
+        -----
+        cube           : Iris cube of gridded data.
+        sites          : Dictionary of site data, including lat/lon and
+                         altitude information.
+        relative_z     : Sets the preferred vertical displacement of the grid
+                         point relative to the site; above/below/None.
+        land_constraint: A boolean that determines if land sites should only
+                         select from grid points also over land.
+        ancillary_data : A dictionary containing additional model data that
+                         is needed.
+                         Must contain {'orography': <cube of orography>}.
+                         Needs {'land': <cube of land mask>} if using land
+                         constraint.
+
+        Returns:
+        --------
+        neighbours: Numpy array of grid i,j coordinates that are nearest to
+                    each site coordinate given. Includes height difference
+                    between site and returned grid point.
+
+        '''
+        # Use the default nearest neighbour list as a starting point, and
+        # if for some reason it is missing, recreate the list using the fast
+        # method.
+        if default_neighbours is None:
+            neighbour_list = self.fast_nearest_neighbour(cube, sites,
+                                                         ancillary_data)
+        else:
+            neighbour_list = default_neighbours
+
+        orography = data_from_ancillary(ancillary_data, 'orography')
```

Comment:
I am passing in the ancillary data, but here I am extracting a particular component of it (and extracting .data from the cube). This prevents the function interface from growing larger when using several ancillary files.

I might look at this again (and again and again) but for now, I think I will leave this be until we chat more about it.